app/javascript/src/PaymentGateways/braintree/utils/useBraintreeHostedFields.ts

```diff
@@ -86,6 +84,7 @@ const useBraintreeHostedFields = (
               postalCode: billingAddress.zip,
               countryCodeAlpha2: billingAddress.countryCode
             },
+            collectDeviceData: true,
```

Comment:
I have a few questions:

1. Shouldn't this verification be done by the server? Calling this from a TS file suggests the client sends the verification request directly to Braintree, bypassing our server, is that expected?
2. Shouldn't we add the IP address to the params here?

Now I'll give you my opinion on your questions to Braintree:

> 1) Is the current integration compliant with the upcoming changes, or do we need to collect more data from the customers at the moment they enter Credit Card details? (e.g. Browser IP address)
> ====> My doubt is about whose IP should be submitted - the card owner's (when they enter the card details through the browser), or our servers (that actually charge the card).

I would say: for sure, send the browser IP when the client adds their CC credentials. However, I would send our server IP when we do recurrent charges without customer being present. I think it's more relevant for them, in order to know who charged what. Also, what's the alternative? storing the browser IP in our DB when they add the CC credentials and then sending that IP again and again at every recurrent charge? That wouldn't make sense, that IP would mean nothing at the moment of a recurrent charge, and it would be not related to the charge at all. So if they fail to answer your question properly, I would send our server IP.

> 2)  Will the payment methods saved until now be used successfully to make payments after August 12?
> ====> I refer to the situation where 3DS was enabled before and the card details have been stored and the payments have been processing successfully - would payments using these existing Payment Methods still work after Aug 12?

> 3) In case the merchant doesn't have 3DS enabled currently, but enables it at a later date, will they need to recreate all the payment methods collecting the required fields?
> ====> What should be the process? Would the end users just need to update/recreate their card details again?

Just to clarify: a merchant here is a 3scale tenant, and the client making the payment is one of their buyers, am I right?

For those two questions I have the same answer: just always send all 3DS fields from now on, it doesn't matter if it was supported in previous charges or not, or whether 3DS is currently supported by the merchant. I don't know what would happen if we send 3DS fields to a merchant that doesn't support them, bu I think those scenarios should be managed by Braintree, not us.

> 4) How can we tell whether the device data and/or the IP address have been collected properly? I can't see these details in the Payment Method (in Vault) even with collectDeviceData: true being passed to the verifyCard() method?
> ====> Does Braintree store this information somewhere?

Doesn't Braintree offers a development environment with fake data? Like Stripe does?

Reply:
Thank you for your comments @jlledom !

I'll try to give answers :)

> 1. Shouldn't this verification be done by the server? Calling this from a TS file suggests the client sends the verification request directly to Braintree, bypassing our server, is that expected?

Yes, this is expected, and this is done in a similar way for all Payment Gateways (PG). The client "talks" directly to the payment gateway, and sends CC data to it, PG returns a nonce that we can then use to send to our server and store what we need to store - typically the expiration date, 4 last digits of the card and the card reference.

I believe this helps ensure that we are PCI compliant, because our server never sees the full CC details (full number and CVV).

Some more info about Braintree Hosted Fields: https://developer.paypal.com/braintree/docs/guides/hosted-fields/overview/javascript/v3

Reply:
> Doesn't Braintree offers a development environment with fake data? Like Stripe does?

It surely does, it's called Sandbox: https://sandbox.braintreegateway.com/

However, I can't see this data in the portal, so I am not sure whether I'm just looking in the wrong place, or they are not stored by Braintree (and just passed directly to Visa or whatever).

Reply:
Thanks for your answers, I have a better picture now.

1. So if the server never gets the actual CC details, how can we perform recurrent charges from the server?

2. Also, I guess we can't collect device data from the server, so are the IP and device data only sent once when manually adding CC details? Or in another words, do we need 3DS fields to perform recurrent charges from server?

Reply:
tl;dr; but some comments.

3DS I think only applies when it is performed. I would guess that recurring payments do not apply.

I see no reason to send server IP because:
* they see the IP request is coming from, so they don't need us to tell them, what's the point to accept a claim when you can see the truth
* we don't know the server IP. It is running inside some VPN with a private IP, this means nothing to the payment processor
* it is not browser IP

To me it only makes sense to send browser IP whenever client initiates a payment so for that particular request, we send IP we received the request from.

Reply:
Yes, @akostadinov I was thinking about it, and I fully agree with you. Plus, also I found a place where Braintree accepts an IP address in the card verification flow, but I don't think there anything like that in the charge request.

Reply:
thanks for the explanation @mayorova